### PR TITLE
Adds volunteer image classification tool, volunteer officer identification tool, admin/moderation capability, leaderboard, and user profiles

### LINKED ISSUE #13: Tagging feature
We want an application that displays submitted images that have not been tagged yet, which one can grab by `select * from officers.raw_images where is_tagged = 'f'`. 

Then this application would allow people to click on a face and drag a box to select the cutout of the face - this is what will be stored in `face_position` in table `faces` and displayed in the digital lineup. Then next to each face, there should be two fields: one for name and one for badge number which the user should fill in one or both of. The user should be able to tag a single image multiple times.

Upon submission of the form, if there is one officer that matches the name or badge number then we add a row in table `faces` linking the `officer_id` with the `img_id` at `face_position`. If there is more than one officer (e.g. in the case of a surname that is very common or a badge number that has been reused) or if there are zero officers (e.g. in the case of a new officer that does not appear in our roster yet) we will need to manually review it.

### LINKED ISSUE #14: Implement `face_position`
I tagged a bunch of photos and manually made cutouts of the officer faces. In the future we should implement `face_position` to make cutouts on the fly of the faces.

### LINKED ISSUE #158: Volunteer leaderboard
It would be nice to have a page that shows the top 10 or 20 users. I think the best metric here would be "Top users by images sorted" but we could also add "Top users by officers tagged".

### LINKED ISSUE #159: User profile pages
People might want to show their OpenOversight profiles to their friends and enemies. This profile should have a route like `/user/<username>` and should contain the person's username, number of images sorted, and number of cops found. A link to the profile should be added to the right side of the navigation toolbar for logged in users.

### PR DESCRIPTION
**Pull request by redshiftzero**
This closes #13 (implement tagging functionality for users), closes #14 (implements `face_position` storage in the database), closes #159 (implement user profiles), and closes #158 (implement a tagging leaderboard). Here is a rundown of the changes:

## `/user`: View user profile page

Each user now has a "Profile" link in the upper right hand of the navigation bar, that takes them to an overview of their OpenOversight activity:

screen shot 2017-01-24 at 12 38 04 am

Users can also look at one another's profiles. When an administrator is moderating, this is a good place to start: check for fradulent classifications / tags by clicking on a user's page and looking at a few of their classifications / tags in order to remove fradulent ones. An administrator can click the "Disable account" to withdraw a user's ability to classify images and add tags. They should only do this if they believe the account is fradulent. The "Admin only" panel only appears if the user is logged in and is an administrator. 

## `/image`: Show image and moderate classifications

The new `/image` route enables a logged in user to view the metadata associated with a particular image, as well as the classification of the image and who it was classified by. An administrator has the option here to change the classification if they decide it is incorrect (again, the "Admin only" panel only appears if the user is logged in and is an administrator):

screen shot 2017-01-24 at 12 38 10 am

Classifying the image as being without police officers will remove it from circulation - it won't be displayed to volunteers to classify or tag. 

## `/tag`: Show tag and moderate tags

The new `/tag` route enables a logged in user to view the officer ID, the image and who it was added by. An administrator has the option here to remove the tag if they decide it is incorrect (again, the "Admin only" panel only appears if the user is logged in and is an administrator):

screen shot 2017-01-24 at 12 38 20 am

## `/leaderboard`: Show top volunteers

This adds a leaderboard:

screen shot 2017-01-24 at 12 38 29 am

Admins can use this to quickly find new active users and can click on the profile to inspect their activity in detail and verify that it is not fradulent. 

## Volunteer Application

Volunteers without administrator privileges will use the new classification and tagging feature. When a user clicks "Volunteer" (formerly "Getting Started"), they will be taken to a page instructing them to login or register:

screen shot 2017-01-24 at 12 35 58 am

Upon login or registering and reloading this page, they will be provided with two options:

screen shot 2017-01-24 at 12 36 16 am

### `/sort` Volunteer image classification task

This task has volunteers sort through submitted (or scraped) images in the database and classify them as being with or without cops:

screen shot 2017-01-24 at 12 37 39 am

As they click on one of the buttons, the classification is saved and another image is displayed. Classifying them as being with cops puts them in the queue to be tagged. Classifying them as being without cops removes them from the queue of images to be sorted and they will never appear in the tagging task.

### `/cop_face` Volunteer tagging task

A user selects the area of the face using the interactive box (this uses the [Cropper JQuery plugin](https://github.com/fengyuanchen/cropper/blob/master/README.md)). A preview of the area selected is displayed on the page. A user then uses the previously implemented tagger search form to find the officer's OpenOversight ID in the roster and enters it in the form. They click add tag. 

screen shot 2017-01-24 at 12 38 48 am

They continuously add tags for every officer in the image. Finally, a user clicks "There are no remaining police officers" to remove this image from the tagging queue.

I need to add more tests and will make a few followup issues for small UX and workflow improvements that should be made probably after 0.2. However with this PR a non-technical user can add officers to the database relatively easily.

### CODE REVIEW

File: OpenOversight/app/templates/profile.html [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+                <td><b>Number of officers found</b></td>
+                <td>{{ user.tags|length }}</td>
+              </tr>
+            </tbody>
+          </table>
+
+          <p><a href="{{ url_for('main.leaderboard') }}"
+                class="btn btn-primary" role="button">Show leaderboard</a></p>
+
+          {% if current_user.is_administrator and user.is_disabled == False %}
```
Comment (freddymartinez9) [inline OpenOversight/app/templates/profile.html]: I agree. We shouldn't ever have disabled admins?

File: OpenOversight/app/templates/tutorial.html [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+          numbers, proceed to Step 2 by clicking "Police officers are clearly
+          visible with badge numbers or names".</p>
+        <p>If you don't see any clear faces of police officers in the image
+          with names or badge numbers, just click
+          "No police officers with badge numbers or names" to advance to the next
+          image. However, if you do then you will proceed to Step 2.
+          </p>
+
+        <h2>Step 2: Map Police Faces to Entries in the Roster</h2>
+        <p>We want you to use the cropper to first select an officer face in the
```
Comment (freddymartinez9) [inline OpenOversight/app/templates/tutorial.html]: is "cropper" the word we want to use?

**Comment by freddymartinez9** [comment]
overall looks good. a few in-line comments.

File: OpenOversight/app/templates/leaderboard.html [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+{% block content %}
+
+<div class="container" role="main">
+
+
+  <div class="page-header">
+    <h1>OpenOversight Volunteer Leaderboard</h1>
+  </div>
+
+  <div class="row">
```
Comment (redshiftzero) [inline OpenOversight/app/templates/leaderboard.html]: Was on the fence about this but in fa48210c0d431ecd62d342bf31144903b0912498 have modified this to require login.

File: OpenOversight/app/main/views.py [hunk 4]
```diff
@@ -30,12 +48,165 @@ def get_officer():
...
+        proper_path = None
+    return render_template('sort.html', image=image, path=proper_path)
+
+
+@main.route('/tutorial')
+def get_tutorial():
+    return render_template('tutorial.html')
+
+
+@main.route('/user/<username>')
```
Comment (redshiftzero) [inline OpenOversight/app/main/views.py]: I figured people might want to show their activity to people that don't have an active OpenOversight account (might help with recruiting), so I left this one as not requiring login. One can't easily get a list of all accounts (could enforce this by modifying the URL here but for the cop threat model I don't think we need to worry about it 😉  - but am happy to revisit if people get concerned about retaliation (e.g. by making the URLs unpredictable or by restricting to login)).

File: OpenOversight/app/templates/profile.html (hunk 1 above)
Comment (redshiftzero) [inline OpenOversight/app/templates/profile.html]: @r4v5: That's a good point - modified in commit c054f0959379f17ea60a57f6d419c3664e59a4e7

@freddymartinez9: Also added an if statement in c054f0959379f17ea60a57f6d419c3664e59a4e7 such that the toggle button is not displayed if the user is an admin (it wouldn't _really_ do anything anyway - their classifications/tags won't get added if their account is disabled but they are administrators so they'd be able to toggle their own account back to enabled).

File: OpenOversight/app/main/views.py [hunk 5]
```diff
@@ -30,12 +48,165 @@ def get_officer():
...
+def profile(username):
+    try:
+        user = User.query.filter_by(username=username).one()
+    except:
+        abort(404)
+    return render_template('profile.html', user=user)
+
+
+@main.route('/user/<username>/<int:toggle>')
```
Comment (redshiftzero) [inline OpenOversight/app/main/views.py]: 🐥  fixed in 560c1a6

File: OpenOversight/app/main/views.py [hunk 6]
```diff
@@ -30,12 +48,165 @@ def get_officer():
...
+    try:
+        tag = Face.query.filter_by(id=tag_id).one()
+        proper_path = serve_image(tag.image.filepath)
+    except:
+        abort(404)
+    return render_template('tag.html', face=tag, path=proper_path)
+
+
+# Rate limiting / CAPTCHA needed on this route
+@main.route('/image/classify/<int:image_id>/<int:contains_cops>')
```
Comment (redshiftzero) [inline OpenOversight/app/main/views.py]: also fixed in 560c1a6

File: OpenOversight/app/main/views.py [hunk 7]
```diff
@@ -30,12 +48,165 @@ def get_officer():
...
+            image.contains_cops = False
+        db.session.commit()
+        flash('Updated image classification')
+    except:
+        flash('Unknown error occurred')
+    return redirect(redirect_url())
+    #return redirect(url_for('main.display_submission', image_id=image_id))
+
+
+@main.route('/tag/delete/<int:tag_id>')
```
Comment (redshiftzero) [inline OpenOversight/app/main/views.py]: also fixed in 560c1a6

File: OpenOversight/app/templates/tutorial.html (hunk 2 above)
Comment (redshiftzero) [inline OpenOversight/app/templates/tutorial.html]: I rewrote the tutorial and added screenshots in commit c96ea74aaa37829d3ecbf6445d25ab54a78ab8cf (the word cropper no longer appears 😜 ).

**Comment by redshiftzero** [comment]
On your input validation note @r4v5: We now should be using IDs (limited to `int`s) everywhere _except_ the profile route - which I've modified in commit b3e9e56 to 404 if any funny business is attempted - specifically if any characters that should not be in the username are passed to the view function.

File: OpenOversight/app/models.py [hunk 8]
```diff
@@ -60,9 +61,16 @@ class Face(db.Model):
     id = db.Column(db.Integer, primary_key=True)
     officer_id = db.Column(db.Integer, db.ForeignKey('officers.id'))
     img_id = db.Column(db.Integer, db.ForeignKey('raw_images.id'))
-    face_position = db.Column(db.String(120), index=True, unique=False, nullable=True)  # No box dtype in SQLalchemy afaict
-    image = db.relationship('Image', backref='face')
+    face_position_x = db.Column(db.Integer, unique=False)
```
Comment (redshiftzero) [inline OpenOversight/app/models.py]: Added the migrations to go from master to this branch in ee7cc70

**Comment by redshiftzero** [comment]
OK made all changes, committed the migrations, and added a bunch of unit tests. This should be good to merge. I might merge this in now but feel free to add further comments here or make issues and I will address them. Thanks for your comments @r4v5 and @freddymartinez9!!

On the staging server, you should be able to just do 

`python manage.py migrate_db` 

and hopefully no bumps occur. I did this in my dev environment (provisioned VM on `master`, checked out `tagging-app` then migrated the database over) to generate the migrations in commit ee7cc70 and nothing exploded.

**Pull request merged**